### PR TITLE
Add simple and searched case when expressions to evaluator

### PR DESCRIPTION
**Pull request by alancai98**
Note: target branch is `feat-value-iters`. Will create a new PR targeting `main` once that branch's PR is merged to `main`.

*Issue #, if available:* None.

*Description of changes:* Adds simple and searched case when expressions to the evaluator. Also fixes projection behavior when a tuple value is `MISSING` (previously included those tuple values; now omits).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-eval/src/eval.rs [hunk 1]
```diff
@@ -326,7 +326,11 @@ impl Evaluable for EvalProject {
             let mut t = Tuple::new();
 
             self.exprs.iter().for_each(|(alias, expr)| {
-                t.insert(alias.as_str(), expr.evaluate(&v_as_tuple, ctx));
+                let evaluated_val = expr.evaluate(&v_as_tuple, ctx);
+                if evaluated_val != Missing {
+                    // Per section 2 of PartiQL spec: "value MISSING may not appear as an attribute value
+                    t.insert(alias.as_str(), evaluated_val);
+                }
```
Comment (alancai98) [inline partiql-eval/src/eval.rs:333]: Previous behavior kept projected tuple values that evaluated to `MISSING`. Changed to now omit those attribute pairs.

Perhaps a better modeling, would be as the spec mentions in [section 6.3](https://partiql.org/assets/PartiQL-Specification.pdf#subsection.6.3) -- that is modeling is a `SELECT VALUE`. This modeling would likely involve changing the logical to evaluation plan in `plan.rs`.

File: partiql-eval/src/plan.rs [hunk 2]
```diff
@@ -187,6 +187,40 @@ impl EvaluatorPlanner {
...
+                            )),
+                            self.plan_values(*case.1),
+                        )
+                    })
+                    .collect();
+                let default = e
+                    .default
+                    .as_ref()
+                    .map(|default| self.plan_values(*default.clone()));
+                // Here, rewrite `SimpleCaseExpr`s as `SearchedCaseExpr`s
```
Comment (alancai98) [inline partiql-eval/src/plan.rs:209]: Modeling simple and searched case expressions separately in the logical plan but reducing simple case expressions to searched case expressions in the evaluator.

File: partiql-eval/src/plan.rs [hunk 3]
```diff
@@ -187,6 +187,40 @@ impl EvaluatorPlanner {
...
+                let cases = e
+                    .cases
+                    .into_iter()
+                    .map(|case| (self.plan_values(*case.0), self.plan_values(*case.1)))
+                    .collect();
+                let default = e
+                    .default
+                    .as_ref()
+                    .map(|default| self.plan_values(*default.clone()));
+                Box::new(EvalSearchedCaseExpr { cases, default })
```
Comment (alancai98) [inline partiql-eval/src/plan.rs:222]: Could also make the `default` a non-option type by following what's mentioned in the SQL-92 spec (section 6.9 <case expression> pg 141):

```
 6) If an <else clause> is not specified, then ELSE NULL is implicit.
```

Comment (alancai98) [reply partiql-eval/src/plan.rs]: Updated in 7de9988dd95e286ba8e7d92198b675f7d2ea4168 to use implicit `ELSE NULL` when no `ELSE` clause is provided.

**Review by jpschorr — APPROVED** [review]

**Comment by alancai98** [comment]
Pretty cool that GitHub changed the target branch (originally targeted `feat-value-iters`, which was merged to `main`), so a separate PR isn't needed.
image

**Pull request merged**